libkbfs/bserver_tlf_journal.go

```diff
@@ -0,0 +1,706 @@
+	return s.writeOrdinalLocked(s.earliestPath(), o)
+}
+
+func (s *bserverTlfJournal) readLatestOrdinalLocked() (journalOrdinal, error) {
+	return s.readOrdinalLocked(s.latestPath())
+}
+
+func (s *bserverTlfJournal) writeLatestOrdinalLocked(o journalOrdinal) error {
+	return s.writeOrdinalLocked(s.latestPath(), o)
+}
+
+const (
+	blockPutOp    = "blockPut"
+	addRefOp      = "addReference"
+	removeRefsOp  = "removeReferences"
+	archiveRefsOp = "archiveReferences"
+)
+
+// A bserverJournalEntry is just the name of the operation and the
+// associated block ID and contexts. Fields are exported only for
+// serialization.
+type bserverJournalEntry struct {
+	// Must be one of the four ops above.
+	Op string
+	ID BlockID
+	// Must have exactly one entry for blockPutOp and addRefOp,
+	// and at least one for removeRefsOp and archiveRefsOp.
+	Contexts []BlockContext
+}
+
+// The functions below are for reading and writing journal entries.
+
+func (s *bserverTlfJournal) readJournalEntryLocked(o journalOrdinal) (
+	bserverJournalEntry, error) {
+	p := s.journalEntryPath(o)
+	buf, err := ioutil.ReadFile(p)
+	if err != nil {
+		return bserverJournalEntry{}, err
+	}
+
+	var e bserverJournalEntry
+	err = s.codec.Decode(buf, &e)
+	if err != nil {
+		return bserverJournalEntry{}, err
+	}
+
+	return e, nil
+}
+
+// readJournalLocked reads the journal and returns a map of all the
+// block references in the journal.
+func (s *bserverTlfJournal) readJournalLocked() (
+	map[BlockID]blockRefMap, error) {
+	refs := make(map[BlockID]blockRefMap)
+
+	first, err := s.readEarliestOrdinalLocked()
+	if os.IsNotExist(err) {
+		return refs, nil
+	} else if err != nil {
+		return nil, err
+	}
+	last, err := s.readLatestOrdinalLocked()
+	if err != nil {
+		return nil, err
+	}
+
+	for i := first; i <= last; i++ {
+		e, err := s.readJournalEntryLocked(i)
+		if err != nil {
+			return nil, err
+		}
+
+		blockRefs := refs[e.ID]
+		if blockRefs == nil {
+			blockRefs = make(blockRefMap)
+			refs[e.ID] = blockRefs
+		}
+
+		switch e.Op {
+		case "put", "addReference":
+			blockRefs.put(e.Contexts[0], liveBlockRef)
+
+		case "removeReferences":
+			for _, context := range e.Contexts {
+				delete(blockRefs, context.GetRefNonce())
+			}
+
+		case "archiveReferences":
+			for _, context := range e.Contexts {
+				blockRefs.put(context, archivedBlockRef)
+			}
+
+		default:
+			return nil, fmt.Errorf("Unknown op %s", e.Op)
+		}
+	}
+	return refs, nil
+}
+
+func (s *bserverTlfJournal) writeJournalEntryLocked(
+	o journalOrdinal, e bserverJournalEntry) error {
+	err := os.MkdirAll(s.journalPath(), 0700)
+	if err != nil {
+		return err
+	}
+
+	p := s.journalEntryPath(o)
+
+	buf, err := s.codec.Encode(e)
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(p, buf, 0600)
+}
+
+func (s *bserverTlfJournal) appendJournalEntryLocked(
+	op string, id BlockID, contexts []BlockContext) error {
+	var next journalOrdinal
+	o, err := s.readLatestOrdinalLocked()
```

Comment:
I don't think so -- I have vague thoughts of a case where you'd have multiple KBFS daemons running on the same machine, but that seems too weird to support.

Added a TODO.